chain-signatures/node/src/protocol/triple.rs

```diff
@@ -346,190 +696,91 @@ impl TripleManager {
-                            // We use Highway Hash because the DefaultHasher doesn't guarantee a consistent output across versions
-                            let entropy =
-                                HighwayHasher::default().hash64(&big_c.to_bytes()) as usize;
-
-                            let num_participants = generator.participants.len();
-                            // This has a *tiny* bias towards lower indexed participants, they're up to (1 + num_participants / u64::MAX)^2 times more likely to be selected
-                            // This is acceptably small that it will likely never result in a biased selection happening
-                            let triple_owner = generator.participants[entropy % num_participants];
-
-                            triple_owner == self.me
-                        };
-
-                        triples.push((triple, triple_is_mine));
-                        if triple_is_mine {
-                            crate::metrics::NUM_TOTAL_HISTORICAL_TRIPLE_GENERATIONS_MINE_SUCCESS
-                                .with_label_values(&[self.my_account_id.as_str()])
-                                .inc();
-                        }
+    pub async fn poke(&self, cfg: &ProtocolConfig) -> Vec<(Participant, TripleMessage)> {
+        let (triples, messages, errors) = {
+            let mut tasks = self.tasks.write().await;
+            tasks
+                .poke(self.me, &self.my_account_id, self.epoch, cfg)
+                .await
+        };
 
-                        // Protocol done, remove it from the ongoing pool.
-                        self.ongoing.remove(id);
-                        self.introduced.remove(id);
-                        // Do not retain the protocol
-                        break false;
-                    }
-                }
+        {
+            let mut gc = self.gc.write().await;
+            for (id, err) in errors.into_iter() {
+                tracing::warn!(id, ?err, "failed to generate triple");
+                gc.insert(id, Instant::now());
             }
-        });
+        }
 
         for (triple, mine) in triples {
             self.insert(triple, mine).await;
         }
 
-        if !errors.is_empty() {
-            tracing::warn!(?errors, "faled to generate some triples");
-        }
-
         messages
     }
+
+    pub fn execute(
+        self,
+        active: &Participants,
+        protocol_cfg: &ProtocolConfig,
+        messages: Arc<RwLock<crate::http_client::MessageQueue>>,
+    ) -> JoinHandle<()> {
+        let active = active.clone();
+        let protocol_cfg = protocol_cfg.clone();
+
+        tokio::task::spawn(async move {
+            if let Err(err) = self.stockpile(&active, &protocol_cfg).await {
+                tracing::warn!(?err, "running: failed to stockpile triples");
+            }
+
+            {
+                let mut messages = messages.write().await;
```

Comment:
This looks like a potential lock-contention again on `messages` if this queue is used by many tasks.

Reply:
good point, the idea is that acquiring the write lock for messages is short lived so actually better if it were an MPSC channel for this purpose, because otherwise we run into this issue if somebody were to try to acquire the message lock somewhere later